### PR TITLE
Implement text-variable-anchor-offset

### PR DESCRIPTION
**Pull request by drwestco**
## Launch Checklist

Consume new text-variable-anchor-offset property from style spec 19.3.0 (https://github.com/maplibre/maplibre-style-spec/issues/112). Add methods to treat text-variable-anchor and text-variable-anchor-offset in a consistent manner. Add render tests (copied/modified from existing text-variable-anchor render tests) to verify consistency.

 - [X] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [X] Briefly describe the changes in this PR.
 - [X] Link to related issues.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [X] Write tests for all new functionality.
 - [ ] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [ ] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

**Comment by HarelM** [comment]
Uh... That will take me a few days to go though... 😀

File: test/integration/render/tests/text-variable-anchor-offset/all-anchors/style.json [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  "layers": [
+    {
+      "id": "text",
+      "type": "symbol",
+      "source": "point",
+      "layout": {
+        "text-field": "x",
+        "text-size": 20,
+        "text-justify": "auto",
+        "text-variable-anchor-offset": [
```
Comment (HarelM) [inline test/integration/render/tests/text-variable-anchor-offset/all-anchors/style.json:95]: The original request was for an icon that is placed above a point and to have the label surround it properly.
I don't see this kind of test here, this is the closest I found.

File: test/examples/variable-offset-label-placement.html [hunk 2]
```diff
@@ -0,0 +1,132 @@
+<!DOCTYPE html>
```
Comment (HarelM) [inline test/examples/variable-offset-label-placement.html:1]: There is a need to add an image to the example, there's a script that generates this image, or you can open this page in the browser and do a screen shot.

File: src/symbol/symbol_layout.ts [hunk 3]
```diff
@@ -789,9 +714,9 @@ function addSymbol(bucket: SymbolBucket,
         useRuntimeCollisionCircles,
         0,
         textBoxScale,
-        textOffset0,
-        textOffset1,
-        collisionCircleDiameter);
+        collisionCircleDiameter,
```
Comment (HarelM) [inline src/symbol/symbol_layout.ts:717]: Looking at the original code, the order of the parameters here is different, is this intended?

File: src/style/style_layer/symbol_style_layer.ts [hunk 4]
```diff
@@ -193,3 +195,162 @@ export function getIconPadding(layout: PossiblyEvaluated<SymbolLayoutProps, Symb
...
+
+// The radial offset is to the edge of the text box
+// In the horizontal direction, the edge of the text box is where glyphs start
+// But in the vertical direction, the glyphs appear to "start" at the baseline
+// We don't actually load baseline data, but we assume an offset of ONE_EM - 17
+// (see "yOffset" in shaping.js)
+const baselineOffset = 7;
+export const INVALID_TEXT_OFFSET = Number.POSITIVE_INFINITY;
+
+export function evaluateVariableOffset(anchor: TextAnchor, offset: [number, number]): [number, number] {
```
Comment (HarelM) [inline src/style/style_layer/symbol_style_layer.ts:221]: Why was this moved here?
If this is a "subject" of its own, I would consider moving it to a separate file all together.

Comment (HarelM) [reply src/style/style_layer/symbol_style_layer.ts:221]: I think overlap mode is somewhat similar in terms of having its own file, doesn't it?

File: src/style/style_layer/symbol_style_layer.ts [hunk 5]
```diff
@@ -193,3 +195,162 @@ export function getIconPadding(layout: PossiblyEvaluated<SymbolLayoutProps, Symb
         values[3] * pixelRatio,
     ];
 }
+
+export enum TextAnchorEnum {
```
Comment (HarelM) [inline src/style/style_layer/symbol_style_layer.ts:199]: Why do we need an enum here? Are we passing this number to the shaders?

File: src/symbol/placement.ts [hunk 6]
```diff
@@ -952,11 +964,12 @@ export class Placement {
...
         if (bucket.hasTextCollisionBoxData()) bucket.textCollisionBox.collisionVertexArray.clear();
 
-        const layout = bucket.layers[0].layout;
+        const layer = bucket.layers[0];
+        const layout = layer.layout;
         const duplicateOpacityState = new JointOpacityState(null, 0, false, false, true);
         const textAllowOverlap = layout.get('text-allow-overlap');
         const iconAllowOverlap = layout.get('icon-allow-overlap');
-        const variablePlacement = layout.get('text-variable-anchor');
+        const hasVariablePlacement = layer._unevaluatedLayout.hasValue('text-variable-anchor') || layer._unevaluatedLayout.hasValue('text-variable-anchor-offset');
```
Comment (HarelM) [inline src/symbol/placement.ts:972]: Why are we using `_unevaluatedLayout` here?

File: src/symbol/placement.ts [hunk 7]
```diff
@@ -561,47 +566,54 @@ export class Placement {
...
-                                textBoxScale, rotateWithMap, pitchWithMap, textPixelRatio, posMatrix,
-                                collisionGroup, overlapMode, symbolInstance, bucket, orientation, variableIconBox, getElevation);
-
-                            if (result) {
-                                placedBox = result.placedGlyphBoxes;
-                                if (placedBox && placedBox.box && placedBox.box.length) {
-                                    placeText = true;
-                                    shift = result.shift;
-                                    break;
+                        let placementPasses = (textOverlapMode === 'never') ? 1 : 2;
```
Comment (HarelM) [inline src/symbol/placement.ts:582]: Are there unit tests I'm missing that are covering this part? If not, I suggest to try and add some, this is clearly something unit tests can check well, with all kinds of edge cases that render tests will probably not cover.

File: src/style/style_layer/symbol_style_layer.ts (hunk 4 above)
Comment (HarelM) [reply src/style/style_layer/symbol_style_layer.ts:221]: All this code also need to be covered with unit tests.

File: src/data/bucket/symbol_attributes.ts [hunk 8]
```diff
@@ -100,8 +100,9 @@ export const symbolInstance = createLayout([
     {type: 'Uint16', name: 'useRuntimeCollisionCircles'},
     {type: 'Uint32', name: 'crossTileID'},
     {type: 'Float32', name: 'textBoxScale'},
-    {type: 'Float32', components: 2, name: 'textOffset'},
     {type: 'Float32', name: 'collisionCircleDiameter'},
+    {type: 'Uint16', name: 'textAnchorOffsetStartIndex'},
```
Comment (HarelM) [inline src/data/bucket/symbol_attributes.ts:104]: Where are those used? is this definition used for serialization to the worker or used for the shaders?

**Comment by HarelM** [comment]
I have more questions than answers unfortunately.
Maybe a short design diagram would help understand how all this works together - which class is passing which data to where and who uses this data.
THANKS!!

File: src/data/bucket/symbol_attributes.ts (hunk 8 above)
Comment (drwestco) [reply src/data/bucket/symbol_attributes.ts:104]: Worker serialization.

File: src/render/draw_symbol.ts [hunk 9]
```diff
@@ -281,7 +279,7 @@ function drawLayerSymbols(
 
     const depthMode = painter.depthModeForSublayer(0, DepthMode.ReadOnly);
 
-    const variablePlacement = layer.layout.get('text-variable-anchor');
+    const hasVariablePlacement = layer._unevaluatedLayout.hasValue('text-variable-anchor') || layer._unevaluatedLayout.hasValue('text-variable-anchor-offset');
```
[the remark this answers is not in the document]
Comment (drwestco) [reply src/render/draw_symbol.ts:282]: This was how the original code was testing for the existence of a property. Depending on how the defaults in the style spec are set, I think using `layout.get` will return something else?
```TypeScript
if (layer._unevaluatedLayout.getValue('text-radial-offset') === undefined) { ... }

[the remark this answers is not in the document]
Comment (drwestco) [reply src/render/draw_symbol.ts:282]: Also, I was keen to avoid creation of the anchor offset collection for a simple existence check, though that may not happen until you actually call `.evaluate`?

File: src/style/style_layer/symbol_style_layer.ts (hunk 5 above)
Comment (drwestco) [reply src/style/style_layer/symbol_style_layer.ts:199]: It gets passed to/from the worker. I didn't see an obvious way to pass the existing anchor values as strings.

File: test/integration/render/tests/text-variable-anchor-offset/all-anchors/style.json (hunk 1 above)
Comment (drwestco) [reply test/integration/render/tests/text-variable-anchor-offset/all-anchors/style.json:95]: Added

File: src/symbol/symbol_layout.ts (hunk 3 above)
Comment (drwestco) [reply src/symbol/symbol_layout.ts:717]: Previously, `text-offset` and `text-radial-offset` was read and converted from EM to PX, and stored in symbol_bucket at lines 605-612. It was stored as two numeric properties - textOffset0, textOffset1. These values were only read back out if `text-variable-anchor` is also in play. Now, since there may be a separate offset per variable anchor, a single pair of numeric properties doesn't suffice. So there's a new serialized array that stores the anchor+offset pairs.

File: src/style/style_layer/symbol_style_layer.ts (hunk 4 above)
Comment (drwestco) [reply src/style/style_layer/symbol_style_layer.ts:221]: Sure. Will move to new file and add unit tests.

File: src/symbol/placement.ts (hunk 6 above)
Comment (drwestco) [reply src/symbol/placement.ts:972]: I think in the case of `text-variable-anchor` and `text-variable-anchor-offset`, using `layout.get` vs `_unevaluatedLayout.hasValue` is equivalent, since neither have a default value and will thus return `undefined`. I ran into existing code that failed to differentiate between an explicit 0 value in the style vs implicit 0 value for an undefined property. Using _unevaluatedLayout seems more reliable here.

File: src/symbol/placement.ts (hunk 7 above)
Comment (drwestco) [reply src/symbol/placement.ts:581]: I don't see any unit tests for placement.ts. Refactoring this class so the internals are exposed for unit testing is well beyond the scope of this PR.

Comment (HarelM) [reply src/symbol/placement.ts:581]: 😔

**Comment by drwestco** [comment]
Added the preview image for the new sample. I think that's all the PR comments addressed?

**Comment by HarelM** [comment]
Change log entry is missing.
I think that it.

**Comment by drwestco** [comment]
> Change log entry is missing. I think that it.

Should I also bump version number in package.json? 3.3.0 maybe?

**Comment by HarelM** [comment]
In a different PR please.

**Review by HarelM — APPROVED** [review]

**Comment by birkskyum** [comment]
Does this close  https://github.com/maplibre/maplibre-style-spec/issues/112 ?

**Comment by HarelM** [comment]
Depends on how we want to track native parity.

**Comment by birkskyum** [comment]
Aaaaah, gotcha. Was there a repro of this feature made for testing that can be converted into a docs example or to present it in the newsletter?

**Comment by HarelM** [comment]
Yes, see the example file in this PR:
https://maplibre.org/maplibre-gl-js/docs/examples/variable-offset-label-placement/
